File: MappedinWithIndoorAtlas/app/src/main/java/com/mappedin/sample/mappedinwithindooratlas/MainActivity.java
package com.mappedin.sample.mappedinwithindooratlas;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import com.indooratlas.android.sdk.IALocation;
import com.indooratlas.android.sdk.IALocationListener;
import com.indooratlas.android.sdk.IALocationManager;
import com.indooratlas.android.sdk.IALocationRequest;
import com.mappedin.sdk.MPIMapView;
import com.mappedin.sdk.listeners.MPIMapViewListener;
import com.mappedin.sdk.models.*;
import com.mappedin.sdk.web.MPIError;
import com.mappedin.sdk.web.MPIOptions;

import java.util.HashMap;
import java.util.Map;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;

public class MainActivity extends AppCompatActivity implements MPIMapViewListener, IALocationListener {

    private static final String TAG = "MainActivity";  //TAG for log statements.

    private final String[] PERMISSIONS = {android.Manifest.permission.ACCESS_FINE_LOCATION,
            android.Manifest.permission.ACCESS_COARSE_LOCATION,
            android.Manifest.permission.INTERNET,
            android.Manifest.permission.BLUETOOTH_SCAN}; //Permissions required for this sample.

    private MPIMapView mMapView;
    private IALocationManager mIALocationManager;

    //Allows looking up the maps by floor. This sample assumes a single building or map group
    //is being used.
    private HashMap<Double, String> mMapsByFloor;

    private boolean mIsBlueDotEnabled = false;  //Monitor the state of Blue Dot visibility.
    private boolean mIsMapLoaded = false; //Monitor whether the map has loaded.

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        setContentView(R.layout.activity_main);

        mMapView = findViewById(R.id.mapView);

        //TODO: Enter your ClientId, ClientSecret and Venue here.
        MPIOptions.Init options = new MPIOptions.Init("5eab30aa91b055001a68e996",
                "RJyRXKcryCMy4erZqqCbuB1NbR66QTGNXVE0x3Pg6oCIlUR1",
                "mappedin-demo-office");

        MPIOptions.ShowVenue showVenueOptions = new MPIOptions.ShowVenue(null,
                true, "#CDCDCD", null);

        mMapView.setListener(this);
        mMapView.loadVenue(options, showVenueOptions, null);

        permissionsCheck();
    }

    //Resume location updates when the app is resumed.
    @Override
    protected void onResume() {
        super.onResume();
        //Enable location updates.
        if (mIALocationManager != null) {
            mIALocationManager.requestLocationUpdates(IALocationRequest.create(), this);
        }
    }

    //Pause location updates when the app is paused.
    @Override
    protected void onPause() {
        super.onPause();
        //Disable location updates.
        if (mIALocationManager != null) {
            mIALocationManager.removeLocationUpdates(this);
        }
    }

    //Tear down the location manager.
    @Override
    protected void onDestroy() {
        mIALocationManager.destroy();
        super.onDestroy();
    }

    //Coordinates requesting permissions if needed or starting location updates if they've been granted.
    private void permissionsCheck() {
        if (arePermissionsGranted()) {
            //Enable Blue Dot
            startShowingLocation();
        } else if (isPermissionRationaleNeeded()) {
            //Show dialog about required permissions.
            showRationaleDialog();
        } else {
            //Ask for permissions.
            requestPermissionLauncher.launch(PERMISSIONS);
        }
    }

    //Checks if the location, Bluetooth scanning and internet permissions have been granted.
    private boolean arePermissionsGranted() {
        for (String permission : PERMISSIONS) {
            if (ContextCompat.checkSelfPermission(this, permission) !=
                    PackageManager.PERMISSION_GRANTED) {
                //A permission has not been granted.
                Log.i(TAG, "Permission is not granted: " + permission);
                return false;
            }
        }
        Log.d(TAG, "All permissions already granted.");
        return true;
    }

    //Checks if the app should explain why it is asking for the permissions it needs.
    private boolean isPermissionRationaleNeeded() {
        for (String permission : PERMISSIONS) {
            if (shouldShowRequestPermissionRationale(permission)) {
                //A permission has not been granted.
                Log.i(TAG, "Show rationale for permission: " + permission);
                return false;
            }
        }
        Log.i(TAG, "Permission rationale not needed.");
        return true;
    }

    //Displays a dialog to the user explaining why the requested permissions are needed.
    private void showRationaleDialog() {
        //Explain to the user why the permissions are required.
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Additional Permissions Needed");
        builder.setMessage("Location permission is required to show your location on the map. Bluetooth scanning is used to discover Bluetooth location beacons.");
        builder.setPositiveButton("Enable", (dialog, id) -> {
            //Request permissions.
            requestPermissionLauncher.launch(PERMISSIONS);
            dialog.dismiss();
        });
        builder.setNegativeButton("Deny", (dialog, id) -> dialog.dismiss());
        builder.create().show();
    }

    // Register the permissions callback, which handles the user's response to the
    // system permissions dialog.
    private final ActivityResultLauncher<String[]> requestPermissionLauncher =
            registerForActivityResult(new ActivityResultContracts.RequestMultiplePermissions(), isGranted -> {
                boolean allGranted = true;
                for (Map.Entry<String, Boolean> pair : isGranted.entrySet()) {
                    if (!pair.getValue()) {
                        allGranted = false;
                        Log.i(TAG, "Permission was not granted: " + pair.getKey());
                    }
                }

                if (allGranted) {
                    //Enable Blue Dot.
                    startShowingLocation();
                } else {
                    //Explain that Blue Dot won't be available.
                    AlertDialog.Builder builder = new AlertDialog.Builder(this);
                    builder.setTitle("Location Disabled");
                    builder.setMessage("Location based permissions are not enabled. Your location will not be shown on the map.");
                    builder.create().show();
                }
            });

    //Enables location updates and display of the user's location on the map as a blue dot.
    private void startShowingLocation() {
        Log.i(TAG, "Enabling Blue Dot.");
        mIsBlueDotEnabled = true;
        mMapView.getBlueDotManager().enable(new MPIOptions.BlueDot());
        mIALocationManager = IALocationManager.create(this);
        mIALocationManager.requestLocationUpdates(IALocationRequest.create(), this);
    }

    @Override
    public void onBlueDotPositionUpdate(@NonNull MPIBlueDotPositionUpdate mpiBlueDotPositionUpdate) {
        Log.d(TAG, "Blue dot position update.");
    }

    @Override
    public void onBlueDotStateChange(@NonNull MPIBlueDotStateChange mpiBlueDotStateChange) {
        Log.d(TAG, "Blue dot state change.");
    }

    @Override
    public void onDataLoaded(@NonNull MPIData mpiData) {
        Log.d(TAG, "Data loaded.");

        mMapsByFloor = new HashMap<>();
        mpiData.getMaps().forEach(mpiMap -> mMapsByFloor.put(mpiMap.getElevation(), mpiMap.getId()));
    }

    @Override
    public void onFirstMapLoaded() {
        Log.d(TAG, "First map loaded.");
        mIsMapLoaded = true;

        //The Mappedin Blue Dot must be enabled after a map is loaded. Location updates may start
        //before the map has been loaded. Enabling it here ensures the Blue Dot is shown.
        if (mIsBlueDotEnabled) {
            mMapView.getBlueDotManager().enable(new MPIOptions.BlueDot());
        }
    }

    @Override
    public void onMapChanged(@NonNull MPIMap mpiMap) {
        Log.d(TAG, "Map changed.");
    }

    @Override
    public void onNothingClicked() {
        Log.d(TAG, "Nothing clicked.");
    }

    @Override
    public void onPolygonClicked(@NonNull MPINavigatable.MPIPolygon mpiPolygon) {
        Log.d(TAG, "Polygon clicked.");
    }

    @Override
    public void onStateChanged(@NonNull MPIState mpiState) {
        Log.d(TAG, "State changed.");
    }

    //Update the user's location on the map as their location changes.
    @Override
    public void onLocationChanged(IALocation iaLocation) {
        MPIPosition.MPICoordinates coordinates = new MPIPosition.MPICoordinates(
                iaLocation.getLatitude(), iaLocation.getLongitude(), iaLocation.getAccuracy(),
                iaLocation.getFloorLevel());
        MPIPosition position = new MPIPosition((double)System.currentTimeMillis(), coordinates,
                "", null);

        double floorLevel = iaLocation.getFloorLevel();
        if (mIsMapLoaded && mMapView.getCurrentMap().getElevation() != floorLevel) {
            Log.i(TAG, "Changing Floor from " + mMapView.getCurrentMap().getElevation() +
                    " to " + floorLevel);
            mMapView.setMap(mMapsByFloor.get(floorLevel), (Function1<? super MPIError, Unit>) null);
        }

        mMapView.getBlueDotManager().updatePosition(position);
        Log.i(TAG, "Location Update Lat: " + iaLocation.getLatitude() +
                " Long: " + iaLocation.getLongitude() +
                " Floor: " + iaLocation.getFloorLevel() +
                " Accuracy: " + iaLocation.getAccuracy());
    }

    //Show location status changes to the user as a toast.
    @Override
    public void onStatusChanged(String s, int status, Bundle bundle) {
        String msg;

        switch (status) {
            case IALocationManager.STATUS_AVAILABLE:
                msg = "Location Status: Available";
                break;
            case IALocationManager.STATUS_LIMITED:
                msg = "Location Status: Limited";
                break;
            case IALocationManager.STATUS_OUT_OF_SERVICE:
                msg = "Location Status: Out of service";
                break;
            case IALocationManager.STATUS_TEMPORARILY_UNAVAILABLE:
                msg = "Location Status: Temporarily unavailable";
                break;
            default:
                msg = "Location Status: Unknown";
                break;
        }

        Log.i(TAG, msg);
        Toast toast = Toast.makeText(getApplicationContext(), msg, Toast.LENGTH_SHORT);
        toast.show();
    }
}

File: MappedinWithOriient/app/src/main/java/com/mappedin/demo/mappedinwithoriient/MainActivity.kt
package com.mappedin.demo.mappedinwithoriient

import android.Manifest
import android.content.DialogInterface
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.mappedin.sdk.MPIMapView
import com.mappedin.sdk.listeners.MPIMapViewListener
import com.mappedin.sdk.models.MPIBlueDotPositionUpdate
import com.mappedin.sdk.models.MPIBlueDotStateChange
import com.mappedin.sdk.models.MPIData
import com.mappedin.sdk.models.MPIHeader
import com.mappedin.sdk.models.MPIMap
import com.mappedin.sdk.models.MPINavigatable
import com.mappedin.sdk.models.MPIPosition
import com.mappedin.sdk.models.MPIState
import com.mappedin.sdk.web.MPIOptions
import me.oriient.ipssdk.api.listeners.IPSCompletionListener
import me.oriient.ipssdk.api.listeners.IPSLoginListener
import me.oriient.ipssdk.api.listeners.IPSPositioningListener
import me.oriient.ipssdk.api.models.IPSError
import me.oriient.ipssdk.api.models.IPSFloor
import me.oriient.ipssdk.api.models.IPSPosition
import me.oriient.ipssdk.api.models.IPSSpace
import me.oriient.ipssdk.ips.IPSCalibrationDialog
import me.oriient.ipssdk.ips.IPSCore
import me.oriient.ipssdk.ips.IPSPositioning
import java.util.function.Consumer

class MainActivity : AppCompatActivity(), MPIMapViewListener, IPSPositioningListener {

    private val TAG = "MainActivity" // TAG for log statements.

    private val PERMISSIONS = arrayOf(
        Manifest.permission.ACCESS_FINE_LOCATION,
        Manifest.permission.ACCESS_COARSE_LOCATION,
        Manifest.permission.INTERNET,
    ) // Permissions required for this sample.

    private lateinit var mMapView: MPIMapView
    private var mIsBlueDotEnabled = false // Monitor the state of Blue Dot visibility.
    private var mIsMapLoaded = false // Monitor whether the map has loaded.

    private lateinit var mMapsByFloor: java.util.HashMap<Int, String>

    private var calibrationDialog: IPSCalibrationDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        mMapView = findViewById<MPIMapView>(R.id.mapView)

        // Use loadVenue to load venue
        // TODO: Enter your ClientId, ClientSecret and Venue here.
        mMapView.loadVenue(
            MPIOptions.Init(
                "YOUR_CLIENT_ID",
                "YOUR_CLIENT_SECRET",
                "YOUR_VENUE_SLUG",
                headers = listOf(MPIHeader("testName", "testValue")),
            ),
            MPIOptions.ShowVenue(labelAllLocationsOnInit = true, backgroundColor = "#CDCDCD"),
        ) {}

        mMapView.listener = this

        permissionsCheck()
    }

    // Coordinates requesting permissions if needed or starting location updates if they've been granted.
    private fun permissionsCheck() {
        if (arePermissionsGranted()) {
            // Enable Blue Dot
            loginToOriient()
        } else if (isPermissionRationaleNeeded()) {
            // Show dialog about required permissions.
            showRationaleDialog()
        } else {
            // Ask for permissions.
            requestPermissionLauncher.launch(PERMISSIONS)
        }
    }

    // Checks if the location, Bluetooth scanning and internet permissions have been granted.
    private fun arePermissionsGranted(): Boolean {
        for (permission in PERMISSIONS) {
            if (ContextCompat.checkSelfPermission(this, permission) !=
                PackageManager.PERMISSION_GRANTED
            ) {
                // A permission has not been granted.
                Log.i(TAG, "Permission is not granted: $permission")
                return false
            }
        }
        Log.d(TAG, "All permissions already granted.")
        return true
    }

    // Checks if the app should explain why it is asking for the permissions it needs.
    private fun isPermissionRationaleNeeded(): Boolean {
        for (permission in PERMISSIONS) {
            if (shouldShowRequestPermissionRationale(permission)) {
                // A permission has not been granted.
                Log.i(TAG, "Show rationale for permission: $permission")
                return false
            }
        }
        Log.i(TAG, "Permission rationale not needed.")
        return true
    }

    // Displays a dialog to the user explaining why the requested permissions are needed.
    private fun showRationaleDialog() {
        // Explain to the user why the permissions are required.
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Additional Permissions Needed")
        builder.setMessage("Location permission is required to show your location on the map. Bluetooth scanning is used to discover Bluetooth location beacons.")
        builder.setPositiveButton("Enable") { dialog: DialogInterface, _: Int ->
            // Request permissions.
            requestPermissionLauncher.launch(PERMISSIONS)
            dialog.dismiss()
        }
        builder.setNegativeButton(
            "Deny",
        ) { dialog: DialogInterface, _: Int -> dialog.dismiss() }
        builder.create().show()
    }

    // Register the permissions callback, which handles the user's response to the
    // system permissions dialog.
    private val requestPermissionLauncher =
        registerForActivityResult(
            ActivityResultContracts.RequestMultiplePermissions(),
        ) { isGranted: Map<String, Boolean> ->
            var allGranted = true
            for ((key, value) in isGranted) {
                if (!value) {
                    allGranted = false
                    Log.i(TAG, "Permission was not granted: $key")
                }
            }
            if (allGranted) {
                // Enable Blue Dot.
                loginToOriient()
            } else {
                // Explain that Blue Dot won't be available.
                val builder =
                    AlertDialog.Builder(this)
                builder.setTitle("Location Disabled")
                builder.setMessage("Location based permissions are not enabled. Your location will not be shown on the map.")
                builder.create().show()
            }
        }

    private fun loginToOriient() {
        Log.i(TAG, "Log in to Oriient.")

        // TODO: Enter your userID, apiKey and ipsDomain here.
        IPSCore.login(
            "YOUR_MAP_ID",
            "YOUR_API_KEY",
            "YOUR_IPS_DOMAIN
            object : IPSLoginListener {
                override fun onError(loginError: IPSError) {
                    Log.d(TAG, "onError() called with error: $loginError")
                }

                override fun onLogin(list: List<IPSSpace>) {
                    Log.d(TAG, "onLogin() called with: list = [$list]")
                    postOriientLogin()
                }
            },
        )
    }

    private fun postOriientLogin() {
        Log.d(TAG, "postOriientLogin")
        // The observer will notify the calibration progress and necessity.
        IPSPositioning.addPositioningListener(this)

        // Turn on automatic calibration.
        IPSPositioning.setAutomaticCalibrationEnabled(true)

        startPositioning()
    }

    private fun startPositioning() {
        // Start positioning.
        IPSPositioning.startPositioning(
            "YOUR_BUILDING_ID",
            null,
            null,
            object : IPSCompletionListener {
                override fun onCompleted() {
                    Log.d(TAG, "Positioning started successfully")
                    // Enable Blue Dot.
                    mMapView.blueDotManager.enable(MPIOptions.BlueDot())
                    mIsBlueDotEnabled = true
                }

                override fun onError(error: IPSError) {
                    Log.e(
                        TAG,
                        "Failed to start positioning: error = " + error.code + " [" + error.message + "]",
                    )
                }
            },
        )
    }

    private fun stopPositioning() {
        IPSPositioning.stopPositioning(object : IPSCompletionListener {
            override fun onCompleted() {
                Log.d(TAG, "Positioning stopped successfully")
                mIsBlueDotEnabled = false
            }

            override fun onError(error: IPSError) {
                Log.e(TAG, "Failed to stop positioning: error = [" + error.message + "]")
            }
        })
    }

    // Start of MPIMapViewListener Methods.
    override fun onBlueDotPositionUpdate(update: MPIBlueDotPositionUpdate) {
    }

    override fun onBlueDotStateChange(stateChange: MPIBlueDotStateChange) {
    }

    override fun onDataLoaded(data: MPIData) {
        mMapsByFloor = HashMap()
        data.maps.forEach(
            Consumer { (id, _, _, elevation): MPIMap ->
                if (elevation != null) {
                    mMapsByFloor[elevation.toInt()] = id
                }
            },
        )
    }

    override fun onFirstMapLoaded() {
        Log.d(TAG, "First map loaded.")
        mIsMapLoaded = true

        // The Mappedin Blue Dot must be enabled after a map is loaded. Location updates may start
        // before the map has been loaded. Enabling it here ensures the Blue Dot is shown.

        // The Mappedin Blue Dot must be enabled after a map is loaded. Location updates may start
        // before the map has been loaded. Enabling it here ensures the Blue Dot is shown.
        if (mIsBlueDotEnabled) {
            mMapView.blueDotManager.enable(MPIOptions.BlueDot())
        }
    }

    override fun onMapChanged(map: MPIMap) {
    }

    override fun onNothingClicked() {
    }

    override fun onPolygonClicked(polygon: MPINavigatable.MPIPolygon) {
    }

    override fun onStateChanged(state: MPIState) {
    }

    // Start IPSPositioningListener methods.
    override fun onError(posError: IPSError) {
        Log.e(TAG, "onLoginError: " + posError.message)
    }

    override fun onPositioningEngineStateChanged(state: Int) {
        when (state) {
            IPSPositioning.PositioningEngineState.CALIBRATING -> {
                Log.i(TAG, "Positioning sate: CALIBRATING")
            }

            IPSPositioning.PositioningEngineState.IDLE -> {
                Log.i(TAG, "Positioning sate: IDLE")
            }

            IPSPositioning.PositioningEngineState.POSITIONING_AND_CALIBRATING -> {
                Log.i(TAG, "Positioning sate: POSITIONING AND CALIBRATING")
            }
        }
    }

    override fun onPositionUpdate(position: IPSPosition) {
        val floorLevel = 3

        Log.i(
            TAG,
            "Location Update Lat: " + position.latitude +
                " Long: " + position.longitude +
                " Floor: " + floorLevel +
                " Accuracy: " + position.accuracy,
        )

        val coordinates = MPIPosition.MPICoordinates(
            position.latitude,
            position.longitude,
            position.accuracy,
            floorLevel,
        )
        val mpiPosition = MPIPosition(
            System.currentTimeMillis().toDouble(),
            coordinates,
            "",
            null,
        )

        mMapView.blueDotManager.updatePosition(mpiPosition)
    }

    override fun onCalibrationProgress(progress: Double) {
        Log.i(TAG, "onCalibrationProgress: $progress")
    }

    @Deprecated("Deprecated in Java")
    override fun onCalibrationGestureNeeded(calibNeeded: Boolean) {
        Log.i(TAG, "onCalibrationGestureNeeded? $calibNeeded")
    }

    override fun onCalibrationGestureNeeded(calibNeeded: Boolean, reason: Int?) {
        Log.i(TAG, "onCalibrationGestureNeeded? $calibNeeded Reason: $reason")

        runOnUiThread {
            if (calibNeeded) {
                if (calibrationDialog == null) {
                    calibrationDialog = IPSCalibrationDialog.Builder(this)
                        .create()
                }
                calibrationDialog?.show()
            } else {
                calibrationDialog?.dismiss()
            }
        }
    }

    override fun onFloorChanged(floor: IPSFloor) {
        Log.i(TAG, "onFloorChanged: $floor")

        // Change maps if the floor has changed.
        if (mIsMapLoaded && mMapView.currentMap?.elevation?.toInt() != floor.order) {
            Log.i(
                TAG,
                "Changing Floor from " + mMapView.currentMap?.elevation +
                    " to " + floor.order,
            )
            mMapsByFloor[floor.order]?.let {
                mMapView.setMap(it) { err ->
                    err?.message?.let { message ->
                        Log.e("setMap ", message)
                    }
                }
            }
        }
    }
}
